src/main/scala/com/fulcrumgenomics/bam/ExtractUmisFromBam.scala

```diff
@@ -0,0 +1,154 @@
+  group = ClpGroups.SamOrBam)
+class ExtractUmisFromBam
+( @arg(flag = "i", doc = "Input BAM file.")                                      val input: PathToBam,
+  @arg(flag = "o", doc = "Output BAM file.")                                     val output: PathToBam,
+  @arg(flag = "r", doc = "The read structure for read 1.")                       val readStructure1: String,
+  @arg(flag = "R", doc = "The read structure for read 2.")                       val readStructure2: String,
+  @arg(flag = "b", doc = "SAM tags in which to store the molecular barcodes (one-per segment) for read 1.")
+                                                                                 val molecularBarcodeTags1: Seq[String] = Seq.empty,
+  @arg(flag = "B", doc = "SAM tags in which to store the molecular barcodes (one-per segment) for read 2.")
+                                                                                 val molecularBarcodeTags2: Seq[String] = Seq.empty,
+  @arg(flag = "a", doc = "Annotate the read names with the molecular barcodes. See usage for more details.") val annotateReadNames: Boolean = true
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+  molecularBarcodeTags1.foreach(tag => if (tag.length != 2) throw new ValidationException("SAM tags must be of length two: " + tag))
+  molecularBarcodeTags2.foreach(tag => if (tag.length != 2) throw new ValidationException("SAM tags must be of length two: " + tag))
+  if (readStructure1.isEmpty) throw new ValidationException("readStructure1 may not be empty")
+  if (readStructure2.isEmpty) throw new ValidationException("readStructure2 may not be empty")
+  molecularBarcodeTags1.foreach { tag => assert(!molecularBarcodeTags2.contains(tag)) }
+  molecularBarcodeTags2.foreach { tag => assert(!molecularBarcodeTags1.contains(tag)) }
+
+  override def execute(): Unit = {
+    val rs1 = ReadStructure(readStructure1)
+    val rs2 = ReadStructure(readStructure2)
+    assert(molecularBarcodeTags1.length == rs1.count {
+      case m: MolecularBarcode => true
+      case _ => false
+    })
+    assert(molecularBarcodeTags2.length == rs2.count {
+      case m: MolecularBarcode => true
+      case _ => false
+    })
+    val in: SamReader = SamReaderFactory.make.open(input.toFile)
+    val iterator: SAMRecordIterator = in.iterator()
+    val out: SAMFileWriter = new SAMFileWriterFactory().makeSAMOrBAMWriter(in.getFileHeader, true, output.toFile)
+    iterator.sliding(2,2).foreach {
+      case Seq(r1: SAMRecord, r2: SAMRecord) =>
+        // verify everything is in order
+        Seq(r1, r2).foreach(r => {
+          if (!r.getReadPairedFlag) throw new IllegalStateException(s"Read ${r.getReadName} was not paired")
+          if (!r1.getReadUnmappedFlag) throw new IllegalStateException(s"Read ${r.getReadName} was mapped")
+        })
+        if (!r1.getFirstOfPairFlag) throw new IllegalStateException(s"Read ${r1.getReadName} was the first end of a pair")
```

Comment:
I think your error messages here and on the next line need "was" -> "was not", right?